### ISSUE TITLE
How to specify queue when using Buffer::builder()?

### ISSUE THREAD

**Issue by NiklasRosenstein** [cogciprocate/ocl#79]
```rust
  let diffuse_buffer = ocl::Buffer::builder()
    .context(pro_que.context())
    .host_data(&diffuse)
    .dims(ocl::SpatialDims::Three(image_dimensions.0 as usize, image_dimensions.1 as usize, 3))
    .build().unwrap();

  let alpha_buffer = ocl::Buffer::builder()
    .context(pro_que.context())
    .host_data(&alpha)
    .dims(ocl::SpatialDims::Two(image_dimensions.0 as usize, image_dimensions.1 as usize))
    .build().unwrap();

  let distance_buffer = ocl::Buffer::<f32>::builder()
    .context(pro_que.context())
    .dims(ocl::SpatialDims::Two(image_dimensions.0 as usize, image_dimensions.1 as usize))
    .build().unwrap();

  // Initialize the distance field.
  println!("Initializing distance buffer...");
  let kernel = pro_que.create_kernel("init_distance_field").unwrap()
    .arg_buf(&distance_buffer)
    .arg_buf(&alpha_buffer)
    .arg_scl(alpha_threshold as u8);
  kernel.enq().unwrap();

  // Read the buffer back into an image.
  println!("Downloading distance buffer...");
  let mut distance_result = vec![0.0f32; distance_buffer.len()];
  distance_buffer.read(&mut distance_result).enq().unwrap();   // << error
```

The error I get is

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: BufferReadCmd::enq: No queue set.', src\libcore\result.rs:860
note: Run with `RUST_BACKTRACE=1` for a backtrace.
make: *** [release-run] Error 101
```

How can I set the queue for the buffer? Using `BufferBuilder::queue()` like this doesn't work

```rust
  let distance_buffer = ocl::Buffer::<f32>::builder()
    .context(pro_que.context())
    .queue(pro_que.queue())
    .dims(ocl::SpatialDims::Two(image_dimensions.0 as usize, image_dimensions.1 as usize))
    .build().unwrap();
```

Because

1. The docs state that we should use *either* of `::context()` or `::queue()`
2. I get a compile error

```rust
error[E0308]: mismatched types
   --> src\main.rs:267:12
    |
267 |     .queue(pro_que.queue())
    |            ^^^^^^^^^^^^^^^ expected struct `ocl::Queue`, found &ocl::Queue
    |
    = note: expected type `ocl::Queue`
               found type `&ocl::Queue`
```

 Thanks!

**Comment by c0gent** [comment]
`pro_que.queue().clone()` :)

**Comment by c0gent** [comment]
Also you can use the [`::queue`](https://docs.rs/ocl/0.14.1/ocl/builders/struct.BufferWriteCmd.html#method.queue) method each time you enqueue a kernel or buffer read/write/etc. (in the method call chain before `.enq()`) if you happen to want to use different queues more conveniently.

**Comment by NiklasRosenstein** [comment]
Thanks for your quick reply! :) That works, only needed to get rid of the `.context()` call then, because it compiled but gave an error at runtime ("assertion failed: self.queue_option.is_none()").

Using `.host_data()` appears to have no effect unfortunately. I also can't use `buffer.write().enq()` when using `.context()` instead of `.queue()`. So now I use `.queue()` instead and `Buffer::write()`. I might have been using the builder wrong ofc.

```rust
  let diffuse_buffer = ocl::Buffer::builder()
    //.context(pro_que.context())
    .queue(pro_que.queue().clone())
    .host_data(&diffuse)
    .dims(ocl::SpatialDims::Three(image_dimensions.0 as usize, image_dimensions.1 as usize, 3))
    .build().unwrap();
  diffuse_buffer.write(&diffuse).enq().unwrap();

  let alpha_buffer = ocl::Buffer::builder()
    //.context(pro_que.context())
    .queue(pro_que.queue().clone())
    .host_data(&alpha)
    .dims(ocl::SpatialDims::Two(image_dimensions.0 as usize, image_dimensions.1 as usize))
    .build().unwrap();
  alpha_buffer.write(&alpha).enq().unwrap();

  let distance_buffer = ocl::Buffer::<f32>::builder()
    //.context(pro_que.context())
    .queue(pro_que.queue().clone())
    .dims(ocl::SpatialDims::Two(image_dimensions.0 as usize, image_dimensions.1 as usize))
    .build().unwrap();
```

**Comment by c0gent** [comment]
> Using .host_data() appears to have no effect unfortunately.

[`host_data`](https://docs.rs/ocl/0.14.1/ocl/builders/struct.BufferBuilder.html#method.host_data) only has any effect when used with the `MEM_USE_HOST_PTR` or `MEM_COPY_HOST_PTR` flags. See the [documentation](https://docs.rs/ocl/0.14.1/ocl/builders/struct.BufferBuilder.html#method.host_data) for the details (and please let me know how I can improve that documentation if necessary). 

> I also can't use buffer.write().enq() when using .context() instead of .queue(). So now I use .queue() instead and Buffer::write().

First of all, a buffer must have a context specified during creation. That is why you must specify either a context or a queue during creation. If you specify a queue, the context associated with that queue is automatically used (which is why it is not allowed to unnecessarily set both).

Next, a buffer can optionally have a default queue. The default queue will be used for all subsequent commands (reads, writes, etc.) if no other queue is specified. That default can be set when creating the buffer or by calling the [`::set_default_queue`](https://docs.rs/ocl/0.14.1/ocl/struct.Buffer.html#method.set_default_queue) method. **If no default queue is set, you must [specify a queue every time you enqueue a command](https://docs.rs/ocl/0.14.1/ocl/builders/struct.BufferCmd.html#method.queue).** On the other hand, if you already have a default queue set and you [specify a queue when enqueuing a command](https://docs.rs/ocl/0.14.1/ocl/builders/struct.BufferCmd.html#method.queue), it will override the default queue if one is set (for that command only).

Basically the whole default queue thing is just a convenience and precludes you from having to specify the queue every time you call a command. If you're using multiple queues, for example: one for reads and one for writes, don't use a default queue and instead specify it for each call. Otherwise, set the default queue and don't worry about.

Hopefully that clarifies a bit.
